meshdb/meshdb.go

```diff
@@ -201,18 +269,154 @@ func (m *MeshDB) FindLatestMiniHeader() (*MiniHeader, error) {
 	return miniHeaders[0], nil
 }
 
+func (m *MeshDB) findEthBackingsWithLessEthPerOrder(target *big.Rat, max int) ([]*ETHBacking, error) {
+	filter := m.ETHBackings.ETHPerOrderIndex.RangeFilter([]byte{}, ratToBytes(target))
+	query := m.ETHBackings.NewQuery(filter).Max(max)
+	var result []*ETHBacking
+	if err := query.Run(&result); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// InsertOrder atomically inserts the given order and updates any relevant ETH
+// backings. It also removes the order with the lowest ETH backing if needed in
+// order to make room. If order storage is full and the given order does not
+// have a high enough ETH backing, it will not be stored (this is not considered
+// an error). InsertOrder will return an error if any corresponding ETH backings
+// cannot be found.
+func (m *MeshDB) InsertOrder(order *Order) error {
+	txn := m.Database.OpenGlobalTransaction()
+	defer func() {
+		_ = txn.Discard()
+	}()
+
+	totalExistingOrders, err := m.Orders.Count()
+	if err != nil {
+		return err
+	}
+	if totalExistingOrders > m.maxOrders {
+		// This should never happen and indicates a bug in meshdb or the db package.
+		// It is hard to fix the problem because we can't efficiently remove more
+		// than one order with the lowest ETH backing per order.
+		return fmt.Errorf("invalid database state: total number of orders (%d) is greater than max orders (%d)", totalExistingOrders, m.maxOrders)
+	}
+
+	var ethBackingForIncomingOrder ETHBacking
+	if err := m.ETHBackings.FindByID(order.SignedOrder.MakerAddress.Bytes(), &ethBackingForIncomingOrder); err != nil {
+		if _, ok := err.(db.NotFoundError); ok {
+			// This should never happen because the ETHWatcher should have already
+			// been notified about this incoming order. We can't fix the problem here
+			// because meshdb doesn't know about the ETHWatcher (if it did it would
+			// violate separation of concerns).
```

Comment:
Sure you're talking about ETHWatcher here? I think you mean `core`'s validation logic (which inserts EthBackings)

Reply:
You're right; this comment should be updated. I wrote it before implementing the validation logic in the `core` package.